posts/2023/installing-farmos-on-ddev-with-composer/index.md

```diff
@@ -0,0 +1,105 @@
+[DDEV](https://ddev.com/) environment may be more straightforward.
+
+> DDEV is an open source tool for running local web development environments for PHP, Python and Node.js, ready in minutes. Its powerful, flexible per-project environment configurations can be extended, version controlled, and shared. DDEV allows development teams to adopt a consistent Docker workflow without the complexities of bespoke configuration.
+
+DDEV's many features are what make it a great way to build Drupal and farmOS
+projects in a local environment:
+
+* Cross-platform, open source software.
+* Supports many versions of Docker or can be used with Colima.
+* When combined with Colima and Mutagen, performance is significantly improved.
+* Automatic HTTPS.
+* Run multiple projects concurrently.
+* Can be used to develop many other CMSs.
+* Support multiple databases (including MySQL, Postgre, SQLite).
+* Lots of developer tools included, and is easy to extend and integrate.  
+
+# Installing farmOS with DDEV and Composer
+
+- First, follow the instructions to install and setup DDEV: https://ddev.readthedocs.io/en/latest/users/install/
+- Once installed, the following commands will create a DDEV project and install farmOS:
+
+```
+ddev config --project-type=drupal10 --docroot=web --create-docroot --database=postgres:15 --php-version=8.1
+ddev get wotnak/ddev-php-geos
+ddev start
+ddev composer create "farmos/project:3.x-dev"
+ddev drush site:install farm --account-name=admin --account-pass=admin -y
+ddev drush uli
+ddev launch
+```
+
+## Usage
+
+### Useful DDEV commands:
+- Start/stop a project: `ddev start`/ `ddev stop`
+- Open project: `ddev launch`
+- Stop ddev: `ddev poweroff`
+- PHP CLI: `ddev php`
+- Additional DDEV commands: https://ddev.readthedocs.io/en/latest/users/usage/cli/
+
+### Drush
+- Drush commands: `ddev drush {drush command}`
+- Login link: `ddev drush uli`
+- CRON: `ddev drush cron`
+- Additional Drush commands: https://www.drush.org/latest/commands/all/
+
+### Composer:
+- Add dependency (exmple `farm_organic`): `ddev composer require drupal/farm_organic`
+- Updating dependencies: `ddev composer update`
+- See farmOS documentation on using Composer: https://farmos.org/hosting/composer/
+
+# Going to production
+
+It should be noted that DDEV _is not_ designed to be used as a production
+hosting environment. DDEV is designed to make managing a local development
+environment easier. Taking a local environment to production depends entirely
+on how that production environment is built. Most composer-based workflows
+will expect the project's `composer.json`, `composer.lock` and any application
+code be checked in to source control, from which the production environment
+can be built.
+
+DDEV provides integrations with hosting providers to simplify the process of
+syncing production data into a local environment. For more information see
+[Hosting Provider Integration](https://ddev.readthedocs.io/en/stable/users/providers/)
```

Comment:
How many folks do you know using DDEV providers to actually deploy farmOS to production?

Unless there's a whole community doing that that I don't know about, I feel like we should be a bit more opinionated here in guiding folks towards the incumbent strategy of building a custom docker/OCI image using composer and deploying that somehow. I think there's more existing knowledge in our community about how to support folks pursuing that strategy.

The providers part still deserves a mention, but not IMHO in a way that makes it seem like that's the recommended strategy at this point.

There could also be a follow-up blog post on using DDEV providers for one or more production deployment scenarios.

Reply:
> The providers part still deserves a mention, but not IMHO in a way that makes it seem like that's the recommended strategy at this point.

Ah I agree, but wasn't trying to recommend hosting providers here, just clarify that DDEV is not meant for anything in production. The DDEV hosting provider instructions mention the common sample integrations, but really this is a feature that can be customized for other production workflows as well. Maybe this can be clarified.

IMO "composer-based workflows" includes the strategy of creating custom docker images. But there could be many other approaches here as well. I'd rather not get too into these details in this post.. maybe we could include a couple links to "Installing farmOS" and "Building farmOS with composer" ?

```suggestion
# Going to production

It should be noted that DDEV _is not_ designed to be used as a production
hosting environment. DDEV is designed to make managing a local development
environment easier. Taking a local environment to production depends entirely
on how that production environment is built. Most composer-based workflows
will expect the project's `composer.json`, `composer.lock` and any application
code be checked in to source control, from which the production environment
can be built. For more information on hosting farmOS see:
- [Installing farmOS](https://farmos.org/hosting/install/)
- [Building farmOS with Composer](https://farmos.org/hosting/composer/)

DDEV offers hosting provider integration that can be customized for different
workflows to sync data between a production and local environment. For more
information and sample integrations see
[Hosting Provider Integration](https://ddev.readthedocs.io/en/stable/users/providers/)
```

Reply:
I think describing the provider integrations as the last paragraph like that - without any caveat about it being somewhat uncharted territory - makes it read a bit like that's the default/recommended path...

How about:

```suggestion
# Going to production

It should be noted that DDEV _is not_ designed to be used as a production
hosting environment. DDEV is designed to make managing a local development
environment easier. Taking a local environment to production depends entirely
on how that production environment is built. Most composer-based workflows
will expect the project's `composer.json`, `composer.lock` and any application
code be checked in to source control, from which the production environment
can be built. For more information on hosting farmOS see:
- [Installing farmOS](https://farmos.org/hosting/install/)
- [Building farmOS with Composer](https://farmos.org/hosting/composer/)

Another strategy that might be worth experimenting with is DDEV's hosting
provider integrations. These can be customized for different workflows to sync
data between a production and the local environment. For more information see
[Hosting Provider Integration](https://ddev.readthedocs.io/en/stable/users/providers/).
```